### PR TITLE
LayerScale on Transolver residuals — cross-dataset stability

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

LayerScale (Touvron et al., 2021 — CaiT) initializes per-channel learnable scalars (α, init=1e-4) on each residual branch output before the skip connection. This makes residual branches near-zero at init, effectively letting the model start as a near-linear function and gradually activate each layer's contribution as training proceeds.

For CFD surrogates with structured physics priors (pressure fields, boundary layers), this should help in two ways:
1. **Stability**: Early training oscillations in Lion-optimized TandemFoil runs (gc needed to prevent divergence) may reflect large residual magnitudes destabilizing the skip connections — LayerScale dampens this automatically.
2. **Depth utilization**: DrivAerML's 4L/512d model may not be fully utilizing all 4 layers at current hyperparameters. LayerScale lets deeper layers activate more selectively.

Paper: https://arxiv.org/abs/2103.17239 (CaiT, Touvron et al. 2021)

## Instructions

Add LayerScale to `target/icml2026/train.py` in the Transolver transformer block. Specifically:

In the `TransolverBlock` (or equivalent residual block), after each sub-layer output (attention + FFN), multiply by a learnable per-channel scalar initialized to `1e-4`:

```python
# In the block __init__:
self.layer_scale_1 = nn.Parameter(1e-4 * torch.ones(hidden_dim))
self.layer_scale_2 = nn.Parameter(1e-4 * torch.ones(hidden_dim))

# In forward, replace:
#   x = x + self.attn(self.norm1(x))
#   x = x + self.ffn(self.norm2(x))
# With:
x = x + self.layer_scale_1.unsqueeze(0).unsqueeze(0) * self.attn(self.norm1(x))
x = x + self.layer_scale_2.unsqueeze(0).unsqueeze(0) * self.ffn(self.norm2(x))
```

Run with `--wandb_group layerscale-cross-dataset` across all 3 datasets using their best known configs:

**Run 1 — TandemFoil (Lion, EMA, gc=0.5):**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 \
  --wandb_group layerscale-cross-dataset
```

**Run 2 — AirfRANS (AdamW, T_max=50):**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 \
  --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --wandb_group layerscale-cross-dataset
```

**Run 3 — DrivAerML (AdamW, 4L/512d, T_max=30):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group layerscale-cross-dataset
```

Distribute across available GPUs. Report val_primary metric for each dataset at best checkpoint.

## Baseline

| Dataset | Metric | Current Best | Best PR |
|---------|--------|-------------|---------|
| TandemFoil | val_primary/surface_pressure_mae | **22.537** | #2924 (Lion lr=1.25e-4, gc=0.5, EMA=0.999) |
| AirfRANS | val_primary/surface_mse | **0.000482** | #2951 (AdamW lr=6e-4, T_max=50) |
| DrivAerML | val_primary/surface_rel_l2_pct | **3.997%** | #2898 (AdamW lr=5e-4, 4L/512d, T_max=30) |

Lower is better on all metrics.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT einar:

## Results

### Summary

LayerScale (per-channel learnable scalars on attention and FFN residual paths) does **not** beat baselines on any of the 3 datasets. On DrivAerML, it causes catastrophic gradient divergence with both init values.

### Metrics at best checkpoint

| Dataset | Metric | init=1e-4 | init=0.1 | Baseline |
|---------|--------|-----------|----------|----------|
| TandemFoil | `val_primary/surface_pressure_mae` | **24.146** (ep154) | **24.508** (ep149) | 22.537 (ep336) |
| AirfRANS | `val_primary/surface_mse` | **0.000892** (ep386) | **0.000637** (ep378) | 0.000482 (ep576) |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **6.326%** (ep176) | **8.250%** (ep144) | 3.997% (ep467) |

### W&B runs

| Experiment | Run ID | Link |
|-----------|--------|------|
| TF init=1e-4 | `cgvnvuwg` | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/cgvnvuwg |
| TF init=0.1 | `8dvk6h07` | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/8dvk6h07 |
| AF init=1e-4 | `gpa13a4p` | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/gpa13a4p |
| AF init=0.1 | `jq2mym4d` | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/jq2mym4d |
| DM init=1e-4 | `la0cys6s` | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/la0cys6s |
| DM init=0.1 | `v8qn1k18` | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/v8qn1k18 |

### Commands used

**init=1e-4:**
```bash
# TandemFoil (GPU 0)
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 \
  --layer-scale-init 1e-4 --wandb-group layerscale-cross-dataset \
  --agent einar --wandb-name "einar/layerscale-tandemfoil"

# AirfRANS (GPU 1)
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --layer-scale-init 1e-4 --wandb-group layerscale-cross-dataset \
  --agent einar --wandb-name "einar/layerscale-airfrans"

# DrivAerML (GPU 2)
CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --layer-scale-init 1e-4 --wandb-group layerscale-cross-dataset \
  --agent einar --wandb-name "einar/layerscale-drivaerml"
```

**init=0.1:**
Same commands on GPUs 3-5 with `--layer-scale-init 0.1`.

### Peak memory

| Dataset | GPU VRAM (from nvidia-smi snapshot) |
|---------|-------------------------------------|
| TandemFoil | ~51 GB |
| AirfRANS | ~40 GB |
| DrivAerML | ~6 GB (low due to diverged state) |

### What happened

**LayerScale did not improve any of the 3 datasets.**

**TandemFoil:** Both init values converge slowly but plateau far from baseline. init=1e-4 slightly outperforms init=0.1 (24.15 vs 24.51). Notably, the init=1e-4 scale parameters go **negative** under Lion optimizer (attn scales reach -0.002, FFN scales stay positive at ~0.025). This suggests Lion's uniform sign-based updates push the scales through zero. Despite this, training remains numerically stable — Lion's gradient clipping keeps the optimization bounded. At ep157, the runs are still converging but at ~7% above baseline with no sign of closing the gap.

**AirfRANS:** init=0.1 outperforms init=1e-4 (0.000637 vs 0.000892), suggesting the larger initial scale helps with AdamW on this dataset. Both are still well above the 0.000482 baseline. Training is stable (grad_norm ~3) and still running at ep386/395, but the gap to baseline suggests LayerScale introduces a capacity ceiling for this architecture.

**DrivAerML — catastrophic divergence:** This is the most striking negative result. **Both** init values cause eventual gradient explosion:
- init=0.1: divergence begins at ep152 (grad_norm jumps from ~34 to 874, then cascades to 10^11+ by ep255). Best metric at ep144 before divergence.
- init=1e-4: delayed divergence begins at ep202 (grad_norm spikes from 0.4 to 55.8, then cascades to 10^7+ by ep260). Best metric at ep176 before divergence.

The delay is proportional to init magnitude — smaller init delays the onset but doesn't prevent it. Current grad_norm values: init=1e-4 at 1.5×10^7, init=0.1 at 9.6×10^11. The DrivAerML runs diverge because the 4-layer/512-dim architecture + AdamW + cosine-T_max=30 cycling creates periodic LR spikes that interact badly with learned residual scales. The LayerScale parameters themselves don't diverge — they stay bounded — but they modulate gradient flow in a way that amplifies the LR cycling instability.

**LayerScale dynamics (from W&B logged per-block means):**
- With Lion (TF): attn scales drift negative (-0.002), FFN scales stay positive (~0.025). Block 0 FFN and Block 1 FFN are the dominant channels.
- With AdamW (AF): scales stay near init values. init=0.1 scales drift slightly — block 2 FFN grows to 0.15, others stay stable.
- With AdamW (DM): init=1e-4 scales go negative (FFN scales: -0.001 to -0.039); init=0.1 scales grow (block 3 FFN reaches 0.42). The scale growth on deeper layers precedes the divergence event.

### Suggested follow-ups

1. **Gradient clipping interaction:** DrivAerML uses grad_clip=0.0 (no clipping) by default. The baseline presumably trains fine without it, but LayerScale may need clipping as a safety net. Testing `--grad-clip 1.0` with LayerScale on DrivAerML could determine if the divergence is purely a gradient scaling issue.

2. **Fixed (non-learnable) LayerScale:** Using a fixed scalar (e.g., 0.1) on residuals without making it learnable would give the initialization benefit without the instability risk from learned parameter drift. This is simpler and would test whether the value of LayerScale is in the initialization or the learning.

3. **Per-dataset tuning vs shared recipe:** The init=0.1 vs init=1e-4 preference flips between datasets (TF prefers 1e-4, AF prefers 0.1), which argues against LayerScale as part of a shared cross-dataset recipe. This contradicts the ICML paper goal of a shared recipe that transfers.

4. **Cosine schedule interaction:** The DrivAerML divergence correlates with cosine LR cycling valleys. A warmup on the LayerScale parameters (freezing them for the first N epochs) or decoupling them from the main optimizer schedule might help, but adds complexity.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3021 because ADVISOR: Closing as a dead end. No run beats any baseline on any dataset, and DrivAerML diverges catastrophically on both init values (max grad norm ~1.45e+09 at init=1e-4, ~1.62e+13 at init=0.1). 

Root causes verified via W&B:
- TandemFoil: Lion optimizer causes LayerScale attention-path scales to drift negative (~-0.002), inverting the residual contribution. Both inits worse than baseline: 23.949 (init=1e-4, +6.3%) and 24.445 (init=0.1, +8.5%).
- AirfRANS: Both inits worse — 0.000753 (init=1e-4, +56%) and 0.000637 (init=0.1, +32%).
- DrivAerML: Diverges at epoch 235 (init=1e-4) and epoch 173 (init=0.1). Cosine LR cycling is the likely trigger — when LR rises after the first valley, LayerScale params amplify updates past stability.

The init-preference flip between datasets is also a flag: a modification that requires per-dataset init tuning to not regress does not provide a portable improvement.

Your diagnostic work was excellent — the per-layer scale tracking and grad-norm analysis correctly identify the failure modes. This rules out the naive LayerScale application. If you want to revisit the concept, the path forward would need to address: (1) LayerScale-aware optimizer (AdaFactor or separate param group with lower LR for scale params), (2) fixed non-learnable scalar (test whether the mechanism provides value without the optimization instability), (3) gradient clipping specifically for scale parameters. But only after one of these shows a clean beat on a single dataset first..
